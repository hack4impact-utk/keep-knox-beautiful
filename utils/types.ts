// Keep these in sync with the backend schema

export interface Volunteer {
    _id?: string;
    email?: string;
    name: string;
    phone?: string;
    totalEvents?: number; // num attended events
    totalHours?: number; // aggregate of attended event hours
    registeredEvents?: (string | Event)[]; // objectid ref is a string
    attendedEvents?: (string | Event)[];
}

export interface Event {
    _id?: string;
    name: string;
    description?: string;
    caption?: string;
    maxVolunteers?: number;
    volunteerCount?: number; // registered + attended
    location?: string;
    groupSignUp?: boolean;
    startDate?: Date;
    endDate?: Date;
    startRegistration?: Date;
    endRegistration?: Date;
    hours?: number;
    image?: ContentfulImage;
    registeredVolunteers?: (string | Volunteer)[]; // objectid ref is a string
    attendedVolunteers?: (string | Volunteer)[];
}

export interface Admin {
    _id: string;
    email: string;
    password?: string;
}

export interface PaginatedVolunteers {
    volunteers: Volunteer[];
    registeredCount: number;
}
export interface LoadMorePaginatedData {
    data: (Event | Volunteer)[];
    isLastPage: boolean;
}

export interface ContentfulImage {
    assetID: string;
    url: string;
}

export class APIError extends Error {
    constructor(public statusCode: number, message: string) {
        super(message);
    }
}

export interface ApiResponse {
    success: boolean;
    message?: string;
    payload?: unknown;
}

// used to keep track for display
// purposes
export interface EventVolunteer {
    volunteer: Volunteer;
    present: boolean;
}
