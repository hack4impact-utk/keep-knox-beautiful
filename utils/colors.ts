export default {
    white: "#FFFFFF",
    primary: {
        purple: {
            300: "#A87B7B",
            500: "#B07B7B",
            700: "#B47B7B",
        },
        orange: {
            300: "#FD9755",
            500: "#FD8033",
            700: "#EC762F",
        },
    },
    secondary: {
        green: {
            300: "#1A934B",
            500: "#1C934B",
            700: "#1F934B",
        },
    },
};
