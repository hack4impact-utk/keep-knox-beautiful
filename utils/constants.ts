export default {
    org: {
        name: {
            full: "Keep Knoxville Beautiful",
            short: "KKB",
        },
        upcomingEventsDesc: "Join us for a workday!",
        images: {
            logo: "logo-transparent.png",
            banner: "banner.png",
            defaultCard: "logo-transparent.png",
        },
        footer: {
            address: "2743B Wimpole Avenue, Knoxville, TN 37914",
            phone: "(865) 521-6957",
            email: "info@Keepknoxvillebeautiful.org",
        },
    },
    revalidate: {
        upcomingEvents: 20,
        eventDesc: 30,
        allVolunteers: 10,
        eventVolunteers: 5,
        volunteerProfile: 10,
    },
    contentfulImageLimit: 20 * 1000 * 1000, // 20 MB
};

/* 
    Some notes:
        - waiver page at /src/pages/waiver.tsx needs to be changed for different orgs
        - update email verification address in env file
        - allow less secure apps: https://myaccount.google.com/lesssecureapps
        - update email template in server/actions/Volunteer.ts in createEmailBody()
*/
