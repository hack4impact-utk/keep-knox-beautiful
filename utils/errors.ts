export default {
    GENERIC_ERROR: "An unexpected error occurred.",
};
