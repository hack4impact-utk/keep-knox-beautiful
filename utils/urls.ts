const prod = process.env.NODE_ENV === "production";

export default {
    baseUrl: prod ? "http://prod_domain_here!" : "http://localhost:3000",
    dbUrl: process.env.MONGO_DB ?? "mongodb://localhost:27017",
    pages: {
        index: "/",
        event: (eventId: string) => `/events/${eventId}`,
        login: "/login",
        adminHome: "/admin",
        events: "/admin/events",
        addEvent: "/admin/events/new",
        updateEvent: (eventId: string) => `/admin/events/${eventId}/update`,
        manageEvent: (eventId: string) => `/admin/events/${eventId}`,
        volunteers: "/admin/volunteers",
        volunteer: (volId: string) => `/admin/volunteers/${volId}`,
    },
    api: {
        events: "/api/events",
        event: (eventId: string) => `/api/events/${eventId}`,
        eventVolunteers: (eventId: string) => `/api/events/${eventId}/volunteers`,
        signup: (eventId: string) => `/api/events/${eventId}/signup`,
        markPresent: (eventId: string, volId: string) => `/api/events/${eventId}/present/${volId}`,
        markNotPresent: (eventId: string, volId: string) => `/api/events/${eventId}/notPresent/${volId}`,
        volunteers: "/api/volunteers",
        volunteer: (volId: string) => `/api/volunteers/${volId}`,
    },
};
