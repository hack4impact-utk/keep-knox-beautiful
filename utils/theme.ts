import createMuiTheme from "@material-ui/core/styles/createMuiTheme";
import colors from "src/components/core/colors";
import { typographyStyles } from "src/components/core/typography/CoreTypography";

const theme = createMuiTheme({
    palette: {
        primary: {
            main: colors.green,
            light: colors.lightGreen,
        },
        secondary: {
            main: colors.blue,
        },
        accent: {
            main: colors.pink,
        },
        background: {
            paper: colors.white,
        },
        text: {
            primary: colors.grays["80"],
            secondary: colors.grays["60"],
            disabled: colors.grays["40"],
        },
    },
    typography: typographyStyles,
});

// custom fields in palette
declare module "@material-ui/core/styles/createPalette" {
    interface Palette {
        accent: Palette["primary"];
    }
    interface PaletteOptions {
        accent: PaletteOptions["primary"];
    }
}

export default theme;
