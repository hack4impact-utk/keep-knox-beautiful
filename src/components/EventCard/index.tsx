// utils
import React, { useState, ReactNode } from "react";
import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";
import { useRouter } from "next/router";

// components
import { Box, Grid } from "@material-ui/core";
import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import IconButton from "@material-ui/core/IconButton";
import Menu from "@material-ui/core/Menu";
import MenuItem from "@material-ui/core/MenuItem";
import CoreTypography from "../core/typography/CoreTypography";

// icons
import MoreVertIcon from "@material-ui/icons/MoreVert";
import RoomOutlinedIcon from "@material-ui/icons/RoomOutlined";
import ScheduleOutlinedIcon from "@material-ui/icons/ScheduleOutlined";

// misc
import { Event } from "utils/types";
import constants from "utils/constants";
import colors from "src/components/core/colors";

// create styles for material-ui
const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        eventCard: {
            height: 360,
            borderRadius: 25,
            // boxShadow: "0 3px 5px -1px rgba(0,0,0,.2),0 5px 8px 0 rgba(0,0,0,.14),0 1px 14px 0 rgba(0,0,0,.12)",
            transition: ".3s ease-out",
            position: "relative",
            cursor: "pointer",
            minWidth: 300,
        },
        thumbnailPlaceholder: {
            background: `#97BBCB url("/${constants.org.images.defaultCard}") no-repeat center`,
            backgroundSize: "100px",
            height: 200,
            transition: ".3s",
            width: "100%",
        },
        hidden: {
            opacity: 0,
        },
        shown: {
            opacity: 1,
        },
        thumbnailOverlay: {
            backgroundColor: "rgba(0,0,0,.5)",
            width: "100%",
            height: "100%",
            position: "relative",
            transition: ".5s",
        },
        eventDate: {
            width: 65,
            height: 65,
            backgroundColor: "#87CD9B",
            position: "absolute",
            right: "5%",
            top: "5%",
            boxShadow: "0px 0px 10px 2px rgba(0, 0, 0, 0.15)",
            borderRadius: "10px",
            color: "white",
            textAlign: "center",
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
        },
        title: {
            fontSize: 22,
        },
    })
);

interface Props {
    event: Event;
    isAdmin?: boolean;
}

interface ThumbProps {
    localHover: boolean;
    children: ReactNode;
}

const EventCard: React.FC<Props> = ({ event, isAdmin = false }) => {
    const classes = useStyles();
    const router = useRouter();

    // display variables for event data
    const eventMonth = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"][
        new Date(event.startDate as Date).getMonth()
    ];
    const eventDate = new Date(event.startDate as Date).getDate();
    const eventLocation = event.location?.split(",")[0];
    const eventStartTime = `${new Date(event.startDate as Date).getHours() % 12 || 12}:${`0${new Date(
        event.startDate as Date
    ).getMinutes()}`.slice(-2)} ${new Date(event.startDate as Date).getHours() >= 11 ? "PM" : "AM"}`;
    const eventEndTime = `${new Date(event.endDate as Date).getHours() % 12 || 12}:${`0${new Date(
        event.endDate as Date
    ).getMinutes()}`.slice(-2)} ${new Date(event.endDate as Date).getHours() > 11 ? "PM" : "AM"}`;

    // state
    const [hover, setHover] = useState(false);
    const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
    const [menuOpen, setMenuOpen] = useState(false);

    const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
        event.preventDefault();
        event.stopPropagation();
        setAnchorEl(event.currentTarget);
        setMenuOpen(true);
        console.log(event.currentTarget, anchorEl);
    };

    const handleClose = () => {
        setAnchorEl(null);
        setMenuOpen(false);
    };

    // helper func to render image or placeholder thumbnail
    const EventThumbnail: React.FC<ThumbProps> = ({ children, localHover = false }) => {
        const hasImage = event.image !== undefined;

        return (
            <div
                className={`${classes.thumbnailPlaceholder}`}
                id="eventThumb"
                style={
                    hasImage
                        ? {
                              height: 200,
                              width: "100%",
                              // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
                              background: `url("${event.image?.url}") no-repeat center`,
                              backgroundSize: "cover",
                          }
                        : {}
                }
            >
                {children}
            </div>
        );
    };

    const handleHover = (event: React.MouseEvent) => {
        setHover(true);
    };

    const handleHoverLeave = (event: React.MouseEvent) => {
        setHover(false);
    };

    return (
        <div>
            <Card
                onMouseEnter={e => handleHover(e)}
                onMouseLeave={e => handleHoverLeave(e)}
                onClick={() => {
                    void router.push(`/events/${event._id as string}`);
                }}
                className={`${classes.eventCard}`}
                elevation={hover ? 20 : 7}
            >
                {/* <CardActionArea> */}
                <EventThumbnail localHover={hover}>
                    <Box className={classes.eventDate}>
                        <span>
                            <CoreTypography variant="h5" style={{ marginBottom: 0, textTransform: "uppercase" }}>
                                {eventMonth}
                            </CoreTypography>
                            <CoreTypography
                                variant="h3"
                                style={{ marginBottom: 0, lineHeight: "20px", fontSize: "1.75em" }}
                            >
                                {eventDate}
                            </CoreTypography>
                        </span>
                    </Box>
                    {isAdmin ? (
                        <div className={classes.thumbnailOverlay} style={hover ? { opacity: 1 } : { opacity: 0 }}>
                            <IconButton
                                aria-label="more options"
                                aria-haspopup="true"
                                onClick={e => handleClick(e)}
                                style={{ position: "absolute", zIndex: 1000, top: "3%" }}
                            >
                                <MoreVertIcon htmlColor="white" />
                            </IconButton>
                        </div>
                    ) : null}
                </EventThumbnail>
                <CardContent>
                    <div style={{ height: 105, overflow: "hidden" }}>
                        <CoreTypography gutterBottom variant="h3" className={classes.title} id="eventName">
                            {event.name}
                        </CoreTypography>
                        <CoreTypography
                            variant="body2"
                            style={{
                                color: "rgba(0,0,0,.4)",
                                fontFamily: "Ubuntu",
                            }}
                            id="eventDesc"
                        >
                            {event.description}
                        </CoreTypography>
                    </div>
                    {/* location and time info */}
                    <Grid container wrap="nowrap" justify="center" alignItems="center" spacing={1}>
                        <RoomOutlinedIcon color="primary"></RoomOutlinedIcon>
                        <Grid item xs={6} style={{ display: "flex", alignItems: "center" }}>
                            <CoreTypography
                                variant="caption"
                                style={{
                                    whiteSpace: "nowrap",
                                    overflow: "hidden",
                                    textOverflow: "ellipsis",
                                    color: colors.pink,
                                    fontWeight: 600,
                                    fontFamily: "Ubuntu",
                                    fontSize: 13,
                                }}
                                id="eventLoc"
                                component="p"
                            >
                                {eventLocation}
                            </CoreTypography>
                        </Grid>
                        <ScheduleOutlinedIcon color="primary"></ScheduleOutlinedIcon>
                        <Grid item xs={5}>
                            <CoreTypography
                                variant="caption"
                                style={{
                                    whiteSpace: "nowrap",
                                    overflow: "hidden",
                                    textOverflow: "ellipsis",
                                    color: colors.pink,
                                    fontWeight: 600,
                                    fontFamily: "Ubuntu",
                                    fontSize: 13,
                                }}
                                id="eventTime"
                                component="p"
                            >
                                {eventStartTime} - {eventEndTime}
                            </CoreTypography>
                        </Grid>
                    </Grid>
                </CardContent>
                {/* </CardActionArea> */}
            </Card>
            <Menu id="simple-menu" anchorEl={anchorEl} keepMounted open={menuOpen}>
                <MenuItem onClick={handleClose}>Edit</MenuItem>
                <MenuItem onClick={handleClose}>Manage</MenuItem>
                <MenuItem onClick={handleClose}>Delete</MenuItem>
            </Menu>
        </div>
    );
};

export default EventCard;
