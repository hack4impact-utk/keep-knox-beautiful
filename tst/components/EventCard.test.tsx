import { shallow } from "enzyme";
import React, { useState } from "react";

import EventCard from "src/components/EventCard";

import { Event } from "utils/types";

const evNoImg: Event = {
    name: "Test Event",
    description: "This is an event description.",
    caption: "This is an event with a test caption an no image.",
    location: "1200 Baker St Knoxville, TN",
    startDate: new Date("3/1/21 12:00 pm"),
    endDate: new Date("3/1/21 2:00 pm"),
    hours: 2,
};

describe("EventCard.tsx", () => {
    it("renders information from event prop correctly", () => {
        const component = shallow(<EventCard event={evNoImg} />);

        const evName = component.find("#eventName");
        const evDesc = component.find("#eventDesc");
        const evLoc = component.find("#eventLoc");
        const evTime = component.find("#eventTime");

        expect(evName.children().text()).toBe(evNoImg.name);
        expect(evDesc.children().text()).toBe("This is an event with a test caption ...");
        expect(evLoc.children().text()).toBe(evNoImg.location?.split(",")[0]);
        expect(evTime.children().reduce((acc, cur) => (acc += cur.text()), "")).toBe("12:00 PM - 2:00 PM");
    });

    // probably unneccessary
    // it("renders placeholder when image is missing", () => {
    //     const component = shallow(<EventCard event={evNoImg} />);

    //     const thumbnail = component.find("#eventThumb");

    //     console.log(thumbnail.getElement().props);
    // });

    // TODO
    it("handles hover", () => {
        const component = shallow(<EventCard event={evNoImg} />);
    });

    it("interacts with admins on hover", () => {
        const component = shallow(<EventCard event={evNoImg} />);
    });

    it("matches snapshot", () => {
        const component = shallow(<EventCard event={evNoImg} />);

        expect(component).toMatchSnapshot();
    });
});
