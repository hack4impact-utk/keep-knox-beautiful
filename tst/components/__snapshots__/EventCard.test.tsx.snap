// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard.tsx matches snapshot 1`] = `
<div>
  <WithStyles(ForwardRef(Card))
    className="makeStyles-eventCard-1"
    elevation={7}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <EventThumbnail
      localHover={false}
    >
      <Styled(MuiBox)
        className="makeStyles-eventDate-6"
      >
        <span>
          <CoreTypography
            style={
              Object {
                "marginBottom": 0,
                "textTransform": "uppercase",
              }
            }
            variant="h5"
          >
            Mar
          </CoreTypography>
          <CoreTypography
            style={
              Object {
                "fontSize": "1.75em",
                "lineHeight": "20px",
                "marginBottom": 0,
              }
            }
            variant="h3"
          >
            1
          </CoreTypography>
        </span>
      </Styled(MuiBox)>
    </EventThumbnail>
    <WithStyles(ForwardRef(CardContent))>
      <div
        style={
          Object {
            "height": 105,
            "overflow": "hidden",
          }
        }
      >
        <CoreTypography
          className="makeStyles-title-7"
          gutterBottom={true}
          id="eventName"
          variant="h3"
        >
          Test Event
        </CoreTypography>
        <CoreTypography
          id="eventDesc"
          style={
            Object {
              "color": "rgba(0,0,0,.4)",
              "fontFamily": "Ubuntu",
            }
          }
          variant="body2"
        >
          This is an event with a test caption ...
        </CoreTypography>
      </div>
      <WithStyles(ForwardRef(Grid))
        alignItems="center"
        container={true}
        justify="center"
        spacing={1}
        wrap="nowrap"
      >
        <Memo([object Object])
          color="primary"
        />
        <WithStyles(ForwardRef(Grid))
          item={true}
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
            }
          }
          xs={6}
        >
          <CoreTypography
            component="p"
            id="eventLoc"
            style={
              Object {
                "color": "#F298AE",
                "fontFamily": "Ubuntu",
                "fontSize": 13,
                "fontWeight": 600,
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
              }
            }
            variant="caption"
          >
            1200 Baker St Knoxville
          </CoreTypography>
        </WithStyles(ForwardRef(Grid))>
        <Memo([object Object])
          color="primary"
        />
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={5}
        >
          <CoreTypography
            component="p"
            id="eventTime"
            style={
              Object {
                "color": "#F298AE",
                "fontFamily": "Ubuntu",
                "fontSize": 13,
                "fontWeight": 600,
                "overflow": "hidden",
                "textOverflow": "ellipsis",
                "whiteSpace": "nowrap",
              }
            }
            variant="caption"
          >
            12:00 PM
             - 
            2:00 PM
          </CoreTypography>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(Card))>
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    id="simple-menu"
    keepMounted={true}
    open={false}
  >
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      Edit
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      Manage
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      Delete
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Menu))>
</div>
`;
